src/ck-ldb/TreeLB.C

```diff
@@ -0,0 +1,537 @@
+  std::fill(expected_incoming.begin(), expected_incoming.end(), 0);
+  std::fill(expected_outgoing.begin(), expected_outgoing.end(), 0);
+  std::fill(load_sent.begin(), load_sent.end(), 0);
+  std::fill(load_received.begin(), load_received.end(), 0);
+  // std::fill(notify_after_transfer.begin(), notify_after_transfer.end(), -1);
+  std::fill(awaitingLB.begin(), awaitingLB.end(), false);
+
+#endif
+}
+
+void TreeLB::configure(json& config)
+{
+#if CMK_LBDB_ON
+
+  LBTreeBuilder* builder;
+  const std::string& tree_type = config["tree"];
+  if (tree_type == "PE_Root")
+  {
+    builder = new PE_Root_Tree();
+  }
+  else if (tree_type == "PE_Process_Root")
+  {
+    builder = new PE_Node_Root_Tree();
+  }
+  else if (tree_type == "PE_Process_ProcessGroup_Root")
+  {
+    builder = new PE_Node_NodeSet_Root_Tree(config["ProcessGroup"]["num_groups"]);
+  }
+  else
+  {
+    CkAbort("TreeLB: configured tree not recognized\n");
+  }
+
+  configure(*builder, config);
+  delete builder;
+
+#endif
+}
+
+void TreeLB::InvokeLB()
+{
+#if CMK_LBDB_ON
+  // NOTE: I'm assuming new LBManager will know when (and when not to) call AtSync
+  if (barrier_before_lb)
+  {
+    contribute(CkCallback(CkReductionTarget(TreeLB, ProcessAtSync), thisProxy));
+  }
+  else
+  {
+    thisProxy[CkMyPe()].ProcessAtSync();
+  }
+#endif
+}
+
+void TreeLB::ProcessAtSync()
+{
+#if CMK_LBDB_ON
+  startTime = CkWallTimer();
+  if (CkMyPe() == 0 && _lb_args.debug() > 0)
+  {
+    CkPrintf("--------- Started LB step %d ---------\n", lbmgr->step());
+  }
+  // CmiAssert(CmiNodeAlive(CkMyPe()));   // TODO move this logic to LBManager
+  int level = 0;  // load balancing starts at the lowest level
+  CkAssert(numLevels > 0 && !awaitingLB[level]);
+  TreeLBMessage* stats = logic[level]->getStats();
+  stats->level = level;
+  awaitingLB[level] = true;
+  sendStatsUp((CkMessage*)stats);
+#endif
+}
+
+// send stats up using the comm-tree for this level
+void TreeLB::sendStatsUp(CkMessage* msg)
+{
+  TreeLBMessage* stats = (TreeLBMessage*)msg;
+  int level = stats->level;
+  int comm_parent_pe = comm_parent[level];
+  // fprintf(stderr, "[%d] TreeLB::sendStatsUp - received msg level=%d comm_parent=%d\n",
+  // CkMyPe(), level, comm_parent_pe);
+  if (comm_parent_pe == -1)
+  {
+    // I'm the root of this comm-tree (current destination of the stats)
+    receiveStats(stats, level);
+  }
+  else if (comm_children[level].size() == 0)
+  {
+    // don't have children so don't have to aggregate any msgs
+    thisProxy[comm_parent_pe].sendStatsUp((CkMessage*)stats);
+  }
+  else
+  {
+    LevelLogic* logic = comm_logic[level];
+    logic->depositStats(stats);
+    if (logic->numStatsReceived() == comm_children[level].size() + 1)
+    {
+      TreeLBMessage* newMsg = logic->mergeStats();
+      newMsg->level = level;
+      thisProxy[comm_parent_pe].sendStatsUp((CkMessage*)newMsg);
+    }
+  }
+}
+
+void TreeLB::receiveStats(TreeLBMessage* stats, int level)
```

Comment:
It's kind of weird and I think inefficient to send a message as a part of a parameter marshalled entry method. I assume the runtime has to copy that message into another message. If you made level the refNum or added another member to the message you could get around that.

Reply:
This (and the other ones you've noted below) should be fine, they aren't entry methods, just regular methods, so there isn't any of this extra copying.